### ISSUE TITLE
[UI] Timeline thumbnails

### ISSUE THREAD

**Issue by elsandosgrande** [olive-editor/olive#506]
It's as the title says. I like that feature in KDEnlive, as it let's me do some things on a frame-by-frame basis. I guess you could do it in Olive as it is right now, but it is significantly harder without a, shall we say, visual aid.

**Comment by BrimsonBhin** [comment]
Would be a cool feature. My only concern is the performance impact of that, so I think it should be optional too

**Comment by alcomposer** [comment]
Yes, I agree the timeline clip view with frames should be an option. Possibly 3:

* Solid, no thumbnails
* Solid, with In/Out thumbnails (at beginning and end of clip) 
* Frame thumbnails

**Comment by elsandosgrande** [comment]
Well, KDEnlive runs well on my laptop with In/Out thumbnails and with frame-by-frame thumbnails at the 12th level of zoom, so I'd say that any dual-core above 2GHz made in the last ~3 years should be fine with this, especially if @itsmattkc can get GPU acceleration to work better than Movit (oh how I hate that library sometimes).

**Comment by alcomposer** [comment]
@elsandosgrande having the option to display the timeline how the user wants to may not be due to system performance, but also personal taste, or even the job at hand.

Having at least these three display options will not only give more options to editors, but also for users with lower performance hardware.

**Comment by elsandosgrande** [comment]
I am not disputing preference, I am just disputing the performance claim (you'd have to be rockin' a really old piece of machinery to have a big impact from that, or something that's between my laptop and a Raspberry Pi; my laptop's a few years old and it was mid-tier when new, so yeah, that's sayin' somethin').

**Comment by itsmattkc** [comment]
There is a potential performance impact of a background thread constantly retrieving frames, but with a low priority thread and enough disk caching I think it could be done with a fairly minimal impact.

**Comment by elsandosgrande** [comment]
Not constantly, but when the clip is added (the thumbnails do not have any effects/overlays added, only the original clip's frames, which should not be too much to render once extracted from the clip). By the way, the frames are only visible in KDEnlive when you're at max zoom (12/12 zoom, 6/12 default).

**Comment by frink** [comment]
Most every pro-grade NLE avoids this because of the performance penalty. You're never going to zoom in to the point that one second takes 2400 pixels - probably not even to the point that one second takes 1200 pixels. If the width of each frame is less than 50px it doesn't make sense rendering that our. And even at 100px that's still a long way from distinguishable change for 1080p60fps. (Note the figures I gave were only for the standard 24fps - 60fps would require 6000px width for 1 sec at 100px and 3000px at 50px...)

For rendering the waveforms one could argue that seeing peaks could help with alignment. While audio sync can show peaks at 1 pixel width you need a full frame to show the accuracy of video. This is why frame-perfect accuracy cannot happen on the timeline without a preview viewport to aid it. Showing a zoomed in waveform preview may be of more aid than showing it in the timeline for synchronization purposes.

The other reason for showing the waveform is to visually predict the amplitude when there is a volume rise so that one can guess where a scene is. I feel that cue labels are a better general indication of that scenario to visually indicate when something happens. As has been discussed above, such a cue set will not require so many background render cycles and produces a better result anyway.

None of the pro-grade NLEs show much pictorial info in the timeline for these reasons...

**Comment by elsandosgrande** [comment]
@frink Look, if you open KDEnlive up and pay with the zoom, you will see that the frame-by-frame thumbnails only show up at max zoom. And speaking of performance, I have not seen any of the penalties you speak of.

**Comment by frink** [comment]
I understand that you would like to be able to jump to positions on the timeline more accurately. But every pro video editor I know just slows the playback speed to find the frame he wants. That way he can preview bigger. If you've ever watched over the shoulder of a pro Lightworks or Avid editor they are too fast with the method mentioned above.

So I guess the question is really whether we are trying to produce something that competes on that level or whether we're just trying to be another Kdenlive, Shotcut prosumer grade or OpenShot, Flowblade hobbyist editor.

My big question isn't so much about what Olive can do the same as other NLEs but what it can do different and better. It's already got the bones of an excellent tweening interface. Clip management is coming along quickly too. If this were easy to do I wouldn't push back but I'm concerned about development time getting stolen from more pressing features like project workflow and color grading.

If most pro editors don't use this feature, (This is just my experience - not a statement of fact...) then I suggest it be pushed back in the line behind imperative pro grade features like decentralized collaboration (open source people are the only ones that know what this really even means...) and workflow management that allows pros to make more money at their job.

I think that has to be our montra if we don't want to become another Pitivi - best intentions but allowing the hobby consumer to define what "professional" is and then shooting right below that mark for release. These projects are perishing for lack of vision into the pro market. But if we're going to do something pro then it's likely to be less friendly for the hobbiest.

**Comment by elsandosgrande** [comment]
@frink Look, I don't have a gun to @itsmattkc 's head of anything, so he'll do it when he has the time to. I am just putting it out there for when that time comes.

**Comment by frink** [comment]
Me neither.

I'm. just concerned that we stay focused on the pro use cases first so that Olive actually hits the mark as a pro editor. I do not believe that this feature is a true need for the pro editor. But I don't care if it's there as log as it doesn't slow me down...

**Comment by unfa** [comment]
@elsandosgrande All Movit managed to do for me is crash Kdenlive... ;)

**Comment by alcomposer** [comment]
@frink considering how pro-active @itsmattkc has been in adding preferences, I think that when this lands it will be an option.

**Comment by xcasxcursex** [comment]
>     * Solid, with In/Out thumbnails (at beginning and end of clip)

Every now and then I check to see if this is in yet. I just realised by the datestamps I've been doing that for two and a half years. I figured nobody would mind me giving this thread a little bump after being that patient ;)

**Comment by xdanielc** [comment]
What can do olive than other editors can't?
Being lightweight, minimal yet feature packed, libre and free and having a recent codebase from the core to the top layer making it rock stable. 
But I also believe this feature to be important. I've used davinci with miniatures everywhere, if you want it you have it, is not just an aesthetic thing, it helps to see that you're pasting the correct color correction settings for instance.
Davinci lags a few tenths of a second when going from cut to cut with up/down arrows, premiere doesn't, I can speculate that's because davinci has heavily styled ui widgets or maybe bc how things are scheduled, maybe it waits until the preview screen can show the project at the current time, idk.
Years ago I used sony vegas on a core 2 duo and felt much more responsive than Davinci on my 9t gen i7, with previews either on or off.

If this gets added, Olive will still be different from the other editors and for some other users it will be a nice shiny feature.
Now if you want a feature for the pro users, cursor snapping on the curve editor might be something that requires more attention imo.

**Comment by xcasxcursex** [comment]
> > ```
> > * Solid, with In/Out thumbnails (at beginning and end of clip)
> > ```
> 
> Every now and then I check to see if this is in yet. I just realised by the datestamps I've been doing that for two and a half years. I figured nobody would mind me giving this thread a little bump after being that patient ;)

OK It's been 3 and a half years.

> In/Out thumbnails (at beginning and end of clip)

I'd settle for just 'in' thumbnails.... just.... something, anything, to visually identify one clip from another. Honestly, if the performance impact of one single read of one single frame of video (or even two if we get super blingy! wow!) and displaying it on the clip is too great, any discussion of focussing on professionals goes out the window. Our PCs can handle a few nanoseconds of read time, but our brains can't handle guessing which one of these 40 blocks on screen has which content.

**Comment by sobotka** [comment]
Which one of you two have been paid for editorial services out of curiosity?

**Comment by xcasxcursex** [comment]
> Which one of you two have been paid for editorial services out of curiosity?

I have. A block of colour is equally (in)discernable from another, regardless of how we pay our bills.

**Comment by itsmattkc** [comment]
> OK It's been 3 and a half years.

Have you actually been following Olive development in that time? If so, you should know that Olive 0.2 was hardly even capable of editing until a few months ago, and as such this was hardly high on the list of priorities.

If you're thinking of Olive 0.1, it's had in-point thumbnails for the entire 3.5 years you've been waiting.

**Comment by xcasxcursex** [comment]
I've been following Olive development of this feature, since it's required for my workflow. In-point thumbnails weren't there when I tried Olive, which is why I subbed to this thread, and show my enthusiasm for this feature once every year or two. Is that a problem?

**Comment by itsmattkc** [comment]
Enthusiasm for a feature is not a problem, but being snarky about it and timing how long it takes to arrive is clearly implying that we're being deliberately negligent. If you know what kind of state Olive 0.2 has been in for the last 3 years, you would know that's not the case.

That being said, you're in luck, because this feature is almost complete in a separate branch and will be merged in soon, so you won't have to return next year.

**Comment by sobotka** [comment]
Wait… I could have sworn both of you chucklefucks made ridiculous appeals to authority of the almighty “professional” nonsense?

> any discussion of focussing on professionals goes out the window.

> Do yourself a favour and take some time off social media so you stop getting the impression that's how real people behave.

More snark.

Here’s an idea, how about accepting ownership of your petty and incessant whinging, and not make the problem about other folks like so often happens with little petty, small dicked open source dweebs?

**Comment by xcasxcursex** [comment]
I wrote a long and polite reply to try and pacify you but honestly I'm tired of swallowing my tongue to try and pacify some asshole who lacks the nads to ever admit they're wrong. It wasn't snark, you're fucking paranoid. I was being friendly from the getgo, smiley faces and all, nothing but enthusiasm for the project and this feature, and you're being a cunt because you just wrongly assumed I was rushing you or some shit, even after I've directly corrected that error. I've made two posts in 3 and a half years that weren't a direct reply to a question directed at me by some paranoid retard and you're crying about incessant posts. Way to turn a hyped supporter into a memory dickhead. Fuck you, I'm out.
